File: gatelogue-types-ts/src/sea.ts
import { Node } from "./node.js";
import { LocatedNode } from "./located.js";

export type SeaMode = "cruise" | "warp ferry" | "traincarts ferry";

export class SeaCompany extends Node {
  get name(): string {
    return this.getColumn("SeaCompany", "name");
  }
  get link(): string | null {
    return this.getColumn("SeaCompany", "link");
  }

  get lines(): SeaLine[] {
    return this.getDerived(SeaLine, "sea/company_lines");
  }
  get stops(): SeaStop[] {
    return this.getDerived(SeaStop, "sea/company_stops");
  }
  get docks(): SeaDock[] {
    return this.getDerived(SeaDock, "sea/company_docks");
  }
}

export class SeaLine extends Node {
  get code(): string {
    return this.getColumn("SeaLine", "code");
  }
  get company(): SeaCompany {
    return this.getColumnFK("SeaLine", "company", SeaCompany)!;
  }
  get name(): string | null {
    return this.getColumn("SeaLine", "name");
  }
  get colour(): string | null {
    return this.getColumn("SeaLine", "colour");
  }
  get mode(): SeaMode | null {
    return this.getColumn("SeaLine", "name");
  }
  get local(): boolean | null {
    const result = this.getColumn<number>("SeaLine", "name");
    return result === null ? null : result === 1;
  }

  get docks(): SeaDock[] {
    return this.getDerived(SeaDock, "sea/line_docks");
  }
  get stops(): SeaStop[] {
    return this.getDerived(SeaStop, "sea/line_stops");
  }
}

export class SeaStop extends LocatedNode {
  get codes(): string[] {
    return this.getSet("SeaStopCodes", "code");
  }
  get company(): SeaCompany {
    return this.getColumnFK("SeaStop", "company", SeaCompany)!;
  }
  get name(): string | null {
    return this.getColumn("SeaStop", "name");
  }

  get docks(): SeaDock[] {
    return this.getDerived(SeaDock, "sea/stop_docks");
  }
  get connectionsFromHere(): SeaConnection[] {
    return this.getDerived(SeaConnection, "sea/stop_connections_from_here");
  }
  get connectionsToHere(): SeaConnection[] {
    return this.getDerived(SeaConnection, "sea/stop_connections_to_here");
  }
  get lines(): SeaLine[] {
    return this.getDerived(SeaLine, "sea/stop_lines");
  }
}

export class SeaDock extends Node {
  get code(): string | null {
    return this.getColumn("SeaDock", "code");
  }
  get stop(): SeaStop {
    return this.getColumnFK("SeaDock", "stop", SeaStop)!;
  }

  get connectionsFromHere(): SeaConnection[] {
    return this.getDerived(SeaConnection, "sea/dock_connections_from_here");
  }
  get connectionsToHere(): SeaConnection[] {
    return this.getDerived(SeaConnection, "sea/dock_connections_to_here");
  }
  get lines(): SeaLine[] {
    return this.getDerived(SeaLine, "sea/dock_lines");
  }
}

export class SeaConnection extends Node {
  get line(): SeaLine {
    return this.getColumnFK("SeaConnection", "line", SeaLine)!;
  }
  get from(): SeaDock {
    return this.getColumnFK("SeaConnection", "from", SeaDock)!;
  }
  get to(): SeaDock {
    return this.getColumnFK("SeaConnection", "to", SeaDock)!;
  }
  get direction(): string | null {
    return this.getColumn("SeaConnection", "direction");
  }
  get duration(): number | null {
    return this.getColumn("SeaConnection", "duration");
  }
}

File: gatelogue-types-rs/src/node/sea.rs
use strum_macros::EnumString;

use crate::{_from_sql_for_enum, _get_column, _get_derived_vec, _get_set, node_type, util::ID};

#[derive(Clone, Copy, PartialEq, Eq, Debug, EnumString)]
pub enum SeaMode {
    #[strum(serialize = "cruise")]
    Cruise,
    #[strum(serialize = "warp ferry")]
    WarpFerry,
    #[strum(serialize = "traincarts ferry")]
    TrainCartsFerry,
}
_from_sql_for_enum!(SeaMode);

node_type!(SeaCompany);
impl SeaCompany {
    _get_column!("SeaCompany", name, String);
    _get_column!("SeaCompany", link, Option<String>);
    _get_derived_vec!(lines, SeaLine, "../sql/sea/company_lines.sql");
    _get_derived_vec!(stops, SeaStop, "../sql/sea/company_stops.sql");
    _get_derived_vec!(
        docks,
        SeaDock,
         "../sql/sea/company_docks.sql"
    );
}

node_type!(SeaLine);
impl SeaLine {
    _get_column!("SeaLine", code, String);
    _get_column!("SeaLine", company, SeaCompany);
    _get_column!("SeaLine", name, Option<String>);
    _get_column!("SeaLine", colour, Option<String>);
    _get_column!("SeaLine", mode, Option<SeaMode>);
    _get_column!("SeaLine", local, Option<bool>);

    _get_derived_vec!(
        docks,
        SeaDock,
        "../sql/sea/line_docks.sql"
    );
    _get_derived_vec!(
        stops,
        SeaStop,
        "../sql/sea/company_stops.sql"
    );
}

node_type!(located SeaStop);
impl SeaStop {
    _get_set!("SeaStopCodes", codes, "code", String);
    _get_column!("SeaStop", company, SeaCompany);
    _get_column!("SeaStop", name, Option<String>);

    _get_derived_vec!(docks, SeaDock, "../sql/sea/stop_docks.sql");
    _get_derived_vec!(
        connections_from_here,
        SeaConnection,
        "../sql/sea/stop_connections_from_here.sql"
    );
    _get_derived_vec!(
        connections_to_here,
        SeaConnection,
        "../sql/sea/stop_connections_to_here.sql"
    );
    _get_derived_vec!(
        lines,
        SeaLine,
        "../sql/sea/stop_lines.sql"
    );
}

node_type!(SeaDock);
impl SeaDock {
    _get_column!("SeaDock", code, Option<String>);
    _get_column!("SeaDock", stop, SeaStop);

    _get_derived_vec!(
        connections_from_here,
        SeaConnection,
        "../sql/sea/dock_connections_from_here.sql"
    );
    _get_derived_vec!(
        connections_to_here,
        SeaConnection,
        "../sql/sea/dock_connections_to_here.sql"
    );
    _get_derived_vec!(
        lines,
        SeaLine,
        "../sql/sea/dock_lines.sql"
    );
}

node_type!(SeaConnection);
impl SeaConnection {
    _get_column!("SeaConnection", line, SeaLine);
    _get_column!("SeaConnection", from, SeaDock);
    _get_column!("SeaConnection", to, SeaDock);
    _get_column!("SeaConnection", direction, Option<String>);
    _get_column!("SeaConnection", duration, Option<u32>);
}
